File: neet_code_all_cpp/graphs/open_the_lock.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <string>
#include <queue>
#include <set>
#include <climits>
#include <cmath>

class Solution {
    public:
        // AC
        int openLock(std::vector<std::string>& deadends, std::string target) {
            std::set<std::vector<int>> deadend_set;
            for(auto& end: deadends) {
                deadend_set.insert(string_to_vec(end));
            }
            if(deadend_set.count(std::vector<int>{0,0,0,0})) {
                return -1;
            }
            auto target_vec = string_to_vec(target);

            std::set<std::vector<int>> seen;
            std::queue<std::vector<int>> q;

            seen.insert(std::vector<int>{0,0,0,0});
            q.push(std::vector<int>{0,0,0,0});

            std::vector<int> dist(10000, INT_MAX);
            dist[0] = 0;

            while(!q.empty()) {
                auto q_front = q.front();
                q.pop();
                auto idx_v = state_to_num(q_front);

                for(int i = 0; i < 4; ++i){
                    auto direction = std::vector{1, -1};
                    for(auto& d: direction) {
                        auto q_front_next = q_front;
                        q_front_next[i] += d;

                        if(q_front_next[i] == -1) {
                            q_front_next[i] = 9;
                        }
                        if(q_front_next[i] == 10) {
                            q_front_next[i] = 0;
                        }

                        if(q_front_next[i] < -1
                                || q_front_next[i] > 10
                                || seen.count(q_front_next)
                                || deadend_set.count(q_front_next)) {
                            continue;
                        }

                        auto idx_next_v = state_to_num(q_front_next);
                        dist[idx_next_v] = std::min(dist[idx_next_v], dist[idx_v] + 1);

                        seen.insert(q_front_next);
                        q.push(q_front_next);
                    }
                }
            }

            auto idx_target = state_to_num(target_vec);
            if(dist[idx_target] == INT_MAX) {
                return -1;
            }

            return dist[idx_target];
        }

    private:
        std::vector<int> string_to_vec(std::string& s) {
            std::vector<int> result;
            for(auto&c: s) {
                result.push_back(c - '0');
            }

            return result;
        }

        int state_to_num(std::vector<int>& state) {
            int num = 0;
            for(int i = 3; i >= 0; --i) {
                num += state[i] * std::pow(10, 3 - i);
            }

            return num;
        }
};

// C#の模範解答より: 合わない...
class SolutionAnsCs {
    public:
        int openLock(std::vector<std::string>& deadends, std::string target) {
            std::unordered_set<std::string> deadend_set;
            for(auto& end: deadends) {
                deadend_set.insert(end);
            }
            if(deadend_set.count("0000") || deadend_set.count(target)) {
                return -1;
            }

            std::queue<std::string> q;
            std::unordered_set<std::string> visit;
            q.push("0000");
            visit.insert("0000");
            int result = 0;
            while(!q.empty()) {
                for(int i = 0; i < (int)q.size(); ++i) {
                    auto current = q.front();
                    q.pop();
                    if(current == target) {
                        return result;
                    }

                    for(auto& s: get_neighbors(current)) {
                        if(!visit.count(s) && !deadend_set.count(s)) {
                            q.push(s);
                            visit.insert(s);
                        }
                    }
                }
                ++result;
            }

            return -1;
        }

        std::vector<std::string> get_neighbors(std::string s) {
            std::vector<std::string> result;
            for(int i = 0; i < (int)s.size(); ++i) {
                auto s_1 = s;
                auto s_2 = s;

                if(s_1[i] == '9') {
                    s_1[i] = '0';
                } else {
                    ++s_1[i];
                }
                result.push_back(s_1);

                if(s_2[i] == '0') {
                    s_2[i] = '9';
                } else {
                    --s_2[i];
                }
                result.push_back(s_2);
            }

            return result;
        }
};

int main(void) {
    using Case = std::pair<std::vector<std::string>, std::string>;
    Case case_1 = {{"0201","0101","0102","1212","2002"}, "0202"};
    // => 6
    Case case_2 = {{"8888"}, "0009"};
    // => 1
    Case case_3 = {{"8887","8889","8878","8898","8788","8988","7888","9888"}, "8888"};
    // => -1

    Solution s_1;
    std::cout << s_1.openLock(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.openLock(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.openLock(case_3.first, case_3.second) << std::endl;

    /*
    SolutionAnsCs s_ans_cs;
    std::cout << s_ans_cs.openLock(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans_cs.openLock(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans_cs.openLock(case_3.first, case_3.second) << std::endl;
    */
}

File: neet_code_all/src/graphs/open_the_lock.rs
use std::collections::{HashSet, VecDeque};

struct Solution {}
impl Solution {
    // AC
    pub fn open_lock(deadends: Vec<String>, target: String) -> i32 {
        let mut deadend_set: HashSet<Vec<i32>> = HashSet::new();
        for end in deadends {
            deadend_set.insert(end.chars().map(|c| c as i32 - '0' as i32).collect());
        }

        if deadend_set.contains(&vec![0, 0, 0, 0]) {
            return -1;
        }

        let target = target
            .chars()
            .map(|c| c as i32 - '0' as i32)
            .collect::<Vec<i32>>();

        let mut seen = HashSet::new();
        let mut q = VecDeque::new();

        seen.insert(vec![0, 0, 0, 0]);
        q.push_back(vec![0, 0, 0, 0]);

        let mut dist = vec![i32::MAX; 10000];
        dist[0] = 0;
        while let Some(q_front) = q.pop_front() {
            let q_front = q_front;
            let idx_v = Self::state_to_num(&q_front) as usize;

            for i in 0..4 {
                for d in [1, -1] {
                    let mut q_front_next = q_front.clone();
                    q_front_next[i] += d;

                    if q_front_next[i] == -1 {
                        q_front_next[i] = 9;
                    }
                    if q_front_next[i] == 10 {
                        q_front_next[i] = 0;
                    }

                    if q_front_next[i] < -1
                        || q_front_next[i] > 10
                        || seen.contains(&q_front_next)
                        || deadend_set.contains(&q_front_next)
                    {
                        continue;
                    }

                    let idx_next_v = Self::state_to_num(&q_front_next) as usize;
                    dist[idx_next_v] = dist[idx_next_v].min(dist[idx_v] + 1);

                    seen.insert(q_front_next.clone());
                    q.push_back(q_front_next.clone());
                }
            }
        }

        let idx_tareget = Self::state_to_num(&target) as usize;
        if dist[idx_tareget] == i32::MAX {
            return -1;
        }

        dist[idx_tareget]
    }

    fn state_to_num(state: &[i32]) -> i32 {
        let n = state.len() - 1;
        let mut num = 0;

        for (i, d) in state.iter().enumerate().rev() {
            num += d * 10_i32.pow((n - i) as u32);
        }

        num
    }
}

fn main() {
    let case_1 = (
        vec!["0201", "0101", "0102", "1212", "2002"]
            .iter()
            .map(|s| s.to_string())
            .collect::<Vec<String>>(),
        "0202".to_string(),
    );
    // => 6
    let case_2 = (
        vec!["8888"]
            .iter()
            .map(|s| s.to_string())
            .collect::<Vec<String>>(),
        "0009".to_string(),
    );
    // => 1
    let case_3 = (
        [
            "8887", "8889", "8878", "8898", "8788", "8988", "7888", "9888",
        ]
        .iter()
        .map(|s| s.to_string())
        .collect::<Vec<String>>(),
        "8888".to_string(),
    );
    // => -1

    println!(
        "case_1: {}",
        Solution::open_lock(case_1.0.clone(), case_1.1.clone())
    );
    println!(
        "case_2: {}",
        Solution::open_lock(case_2.0.clone(), case_2.1.clone())
    );
    println!(
        "case_3: {}",
        Solution::open_lock(case_3.0.clone(), case_3.1.clone())
    );
}
